### ISSUE TITLE
Memory Management functions don't work across shared library borders

### ISSUE THREAD

**Issue by RolfKal** [open62541/open62541#6843]
## Description
When compiling open62541 as shared library, there is a problem when using this in another application that is not compiled with the same or similar compiler toolchain.

## Background Information / Reproduction Steps
The functions UA_malloc(), UA_calloc(), UA_realloc() and UA_free() are simply defines for the C runtime malloc family. This does not allow using any of these functions for memory allocation for elements in UA data structures that are later managed by calls to UA_clear or UA_delete for the entire structure, if the calling application is not compiled/linked with the same C compiler than the open62541 shared library, since the malloc() used in the calling application may be from a different C runtime than the free() used in the open62541 shared library.

This is for instance a problem when trying to use UA_ReadRequest with multiple nodes in one request. This requires allocation of an array of UA_ReadValueId's but there is only a UA_ReadValueId_new() function which allocates a single element. So I tried to use UA_malloc() but got fatal crashes when in the end trying to call UA_ReadRequest_clear() to clean up the request, since the UA_ReadValueId array was not allocated from the same heap.

It may be possible to simply export a UA_newArray() function and allow implementation of the relevant arrays through inline macros like for the new, clear and delete functions.

It may be also useful to actually export the UA_malloc(), UA_calloc(), UA_realloc() and UA_free() as real functions when compiling as shared library, so that it can be used to allocate memory for UA data structures from applications calling the open62541 library as shared library.

Used CMake options:

BUILD_SHARED_LIBS

## Checklist

Please provide the following information:

 - [x] open62541 Version (release number or git tag): 1.4.6
 - [ ] Other OPC UA SDKs used (client or server): 
 - [x] Operating system: Windows 10
 - [ ] Logs (with `UA_LOGLEVEL` set as low as necessary) attached
 - [ ] Wireshark network dump attached
 - [ ] Self-contained code example attached
 - [ ] Critical issue

**Comment by RolfKal** [comment]
Figured out that there is an UA_Array_new() function which solves most of these problems. Still use of UA_malloc() and friends in a calling application is absolutely not advisable when using open62541 as shared library.

Would a patch to add some extra inline macros such as UA_ReadValueId_newArray() based on UA_Array_new() be welcome?

**Comment by jpfr** [comment]
> The functions UA_malloc(), UA_calloc(), UA_realloc() and UA_free() are simply defines for the C runtime malloc family. This does not allow using any of these functions for memory allocation for elements in UA data structures that are later managed by calls to UA_clear or UA_delete for the entire structure, if the calling application is not compiled/linked with the same C compiler than the open62541 shared library, since the malloc() used in the calling application may be from a different C runtime than the free() used in the open62541 shared library.

`UA_malloc` et al are macros that you can override in the `config.h`.
The macro can also bet set to point to a (global) variable with a function-pointer.

This gives maximum flexibility. What are you missing?

**Comment by jpfr** [comment]
I don't understand. `UA_Array_new` uses `UA_calloc` internally.
`UA_calloc` and `UA_malloc` are the same when it comes to heap usage.
That does not change anything.

Let me expand on the solution I mentioned above, using a global singleton variable.

```c
// when compiling the library:
#include <malloc.h>
void * (*myMalloc)(size_t) = malloc; // a global variable pointing to malloc

#define UA_MALLOC myMalloc // make UA_MALLOC resolve to the global variable

// in the code using the library
#include <my_arena_heap.h>
myMalloc = myArenaMalloc; // override the global variable with a different malloc implementation
```

**Comment by RolfKal** [comment]
It's not about using UA_Alloc or UA_calloc! When using UA_Array_new() in an external application I'm using an export from the open62541.dll. This internally calls UA_calloc which will call calloc and eventually maybe malloc. But these malloc and calloc are from the C runtime library that was linked with the open62541.dll. If I compile that DLL in Visual Studio 2022 for instance, it will link to MSVCRT140.dll and use malloc and calloc from there.

If I compile the calling application that uses this open62541.dll with lets say MinGW and use UA_malloc in that application to allocate an array that I then assign to a structure passed to some open62541 APIs, and then consequently call UA_something_clear or UA_something_delete this will eventually run across this internal pointer and attempt to deallocate it with UA_free, which links to the free function in the C runtime linked into the open62541.dll. But my MinGW compiled application uses a different C runtime library with different malloc and friends functions. The UA_malloc in my MinGW application mallocs the memory from the MinGW C runtime library heap, but the UA_free in the open624541.dll that is eventually called through UA_something_clear will be a call to the free function in the MSVCRT140.dll. They are working on completely independent heaps and so the free in MSVCRT140.dll will simply crash on that pointer that was created in the MingW C runtime library.

It's unpractical to expect that a DLL is always called from another binary that is compiled in the same C compiler than the DLL. Before Visual Studio 2015 it was even so that every Visual Studio version had its own C runtime library and you could not share C runtime objects across DLL boundaries unless caller and callee were compiled in exactly the same version. Sharing C runtime objects such as file ids or memory blocks across DLL boundaries is anyhow newer a very good idea.

**Comment by jpfr** [comment]
I asked two things:
1. How is using `UA_Array_new` any better than UA_malloc directly?
2. If the different heap implementations are a problem, why don't you use the global singleton approach?

Your response answered neither.

By the way, also check the definition of `UA_INLINABLE`.
You can compile in a way that UA_ReadValueId_new is compiled into the .dll and not a macro pointing to `UA_calloc`.

You can reopen if there is a good argument that neither of the above can be used as a workaround.

**Comment by RolfKal** [comment]
> 1. How is using UA_Array_new any better than UA_malloc directly?

I tried to explain:

- UA_Array_new is a normally exported function and operates on the C runtime heap in the DLL, so uses the right heap manager. 
- It solves the problem of making sure the memory is allocated from the same heap that it will eventually be freed from.
- It also is the logical function to deal with arrays of UA datatypes. Using UA_malloc for that is actually kind of ugly, considering the whole existing UA_TYPES datatype description system, which takes care of properly defining the various types.
- It does not require to customize the open62541 library itself in any way with some extra initialization function to work. Your proposal would require to add some initialization function to the DLL, but my intention is to keep the open62541 library as standard as possible, so a user of my application can (within reason) later go and recompile the DLL from newer code and get possible bug fixes without me having to recompile my application every time and without me having to explain that user how to patch his new sources to work with my customized DLL interface.

My only question at this point really is if it would be useful to add extra inline functions such as UA_ReadValueId_Array_new(size_t size) in addition to the already existing UA_ReadValueId_new() function. There are other similar structures that a user of the open62541 library may want to allocate as array rather than a single element pointer.

> You can compile in a way that UA_ReadValueId_new is compiled into the .dll and not a macro pointing to UA_calloc.

- When I checked it called UA_new(&UA_TYPES[UA_TYPES_READVALUEID]) and since UA_new is an exported function this is not a problem. 
- But it only allocates a single element of ReadValueID, I want to allocate an array of them to combine several value requests in a single service request.
